### ISSUE TITLE
Multiple matching rules causes unwanted window movement

### ISSUE THREAD

**Issue by overhacked** [koekeishiya/yabai#1383]
When multiple rules match a window, yabai documentation says that values are applied in order and the last rule wins, but the settings from each rule in the rule list are applied to the window before the last rule is reached. The result is that windows that are managed by default or set `managed=on` in an early rule, then `managed=off` in a later rule, will snap into the layout, then pop back out to their floating position. This also causes all the other windows in the layout to briefly resize and reposition.

This is especially a problem for "blanket" rules that generically match certain kinds of windows, for example windows with a blank title (`title=^$`), meant to override application-specific rules that appear earlier in the rule list. The "no-title" window example includes the tiny auto-correct suggestions that Mail.app and other text editing apps pop over the cursor position. When the jumble of incomplete rule application occurs, these pop-over, tiny windows disappear instead of returning to their original position at the cursor.

**Comment by koekeishiya** [comment]
> but the settings from each rule in the rule list are applied to the window before the last rule is reached. 

This is intentional, just badly worded. I'll look into it at some point in the future.

**Comment by overhacked** [comment]
The issue I'm observing is mostly with the `managed` property. If an earlier rule sets a window `managed=on` and a later rule sets `managed=off`, then the whole layout "flickers" as the window pops in and out while rules are being applied. Would you consider the approach in PR #1384 of synthesizing all the rules before actually applying them to the window?

**Comment by koekeishiya** [comment]
There are multiple scenarios in which rules are applied.

1. When a new rule is added at any point in time using the `yabai -m rule --add` command, that rule (and only that rule) should be applied to the matching windows. There is no merging available here as other rules should not be reapplied.

2. When yabai launches, the current system state is detected and cached (applications and windows that are available on the visible spaces). These windows will not have any rules applied at this point, because this is happening before the config runs.  When the config runs, rules are added one by one using the `yabai -m rule --add` command, which triggers an application of said rule for all windows -- this will cause several rules to be applied iteratively to the same window.

3. When a new window is created, all existing rules will be applied to that window. This is the case your PR solves if I read this correctly.

I'm not a huge fan of your solution as it only applies in scenario 3, which causes inconsistencies within the rules system.
I want the behaviour to remain consistent across the board, because everything is configurable during both config-time and runtime.

**Comment by overhacked** [comment]
Just updated the linked PR #1384 that provides a solution for (1) & (2) similar to #1371. It would now be possible to add multiple rules in a single command line, then the resultant set of rules would be reapplied to all windows using the same merging strategy as (3). E.g.:
```
yabai -m rule --add 'app=^Floating$' manage=off --add 'title=^Do not float' manage=on
```

**Comment by koekeishiya** [comment]
Is there a scenario where this is useful that does not include `manage=on`? Because you are using that property in a way that is not / was never intended to be used.

**Comment by overhacked** [comment]
I see your point. I'm using yabai as an "opt-in" window manager, so I create one `app=. manage=off` rule as the first rule, then many `app=^APPNAME$ manage=on` rules for all the apps that I know I want tiled, that work well for me as tiled windows. Maybe I'm going about this wrong way?

I can imagine a set of rules that uses the `display` property that would exhibit similar behavior, though:
```sh
# Put all Mail.app windows on the second display...
yabai -m rule --add app='^Mail$' display=2
# ...except the main message list window on the first display.
yabai -m rule --add app='^Mail$' title=' (messages|unread)$' display=1
```

Whenever the main message window was opened or unminimized, it would pop to display 2 for a few milliseconds before the second rule was evaluated. While it would be possible in this case to make the two rules mutually exclusive, that isn't always practical and leads to a more verbose rule set.

Another example:
```sh
# Make the mouse follow focus for all Photoshop windows
yabai -m rule --add app='^Photoshop$' mouse_follows_focus=on

###
# ... Maybe lots of app-specific rules here ...
###

# Don't make the mouse snap into any window without a title
yabai -m rule --add title='^$' mouse_follows_focus=off
```

So if *any* rule sets `mouse_follows_focus=on` before the later "generic" title rule sets `mouse_follows_focus=off`, then the mouse will snap into the new window before the overriding rule takes effect.

When I read the documentation and started configuring yabai, I understood the order of rules to be a simple priority system where the last matching rule wins if there are conflicting window attributes. Are there any advantages to applying the rules one at-a-time?

**Comment by koekeishiya** [comment]
So the `manage=on` property is specifically there to allow applications that are not detected by yabai by default (due to bad accessibility [role](https://developer.apple.com/documentation/applicationservices/carbon_accessibility/roles?language=objc) and [subrole](https://developer.apple.com/documentation/applicationservices/carbon_accessibility/subroles?language=objc) assignments from the creator, usually happens to cross platform software that is focused on Linux as the main platform), and is not really the inverse of `manage=off` which makes yabai treat a window as floating.

Your first rule to make all applications float is fine, but you should then use the signal system, specifically hook into the `window_created` signal, to perform the filtering you are currently doing with the rule system. Instead of using `manage=on`, you should issue `yabai -m window $YABAI_WINDOW_ID --toggle float`. To retrieve the information you need to decide if a window should be un-floated, you should do a query first: `yabai -m query --windows --window $YABAI_WINDOW_ID`

Something like the following sample. 
You can execute whatever you want in place of *yabai_tile.sh* (python, ruby, compiled binaries), it doesn't need to be written in bash.

yabairc
```sh
yabai -m rule --add app="." manage=off
yabai -m signal --add event=window_created action="~/scripts/yabai_tile.sh"
```

yabai_tile.sh
```sh
#!/usr/bin/env sh

data=$(yabai -m query --windows --window $YABAI_WINDOW_ID)

app=$(echo $data | jq .app)
title=$(echo $data | jq .title)

# if match then
yabai -m window $YABAI_WINDOW_ID --toggle float
```

>. Whenever the main message window was opened or unminimized, it would pop to display 2 for a few milliseconds before the second rule was evaluated. While it would be possible in this case to make the two rules mutually exclusive, that isn't always practical and leads to a more verbose rule set.


That is true, but I think it should be fairly straightforward in most cases.

The other example regarding `mouse_follows_focus` doesn't make much sense to me, as this is only triggered when a window is focused, so I don't see how that is affected by having multiple rules applied iteratively. Maybe I'm missing something subtle here though.

Overall the intention is that rules are meant to be used for fairly simple things, and anything else can be achieved by using the signal system (either on its own, or in combination with a few basic rules).

> Are there any advantages to applying the rules one at-a-time?

Not really; I just prefer the system remain simple. Even with your latest proposal that adds a solution to the previous listed caveats regarding the behaviour, it is still gated behind a new syntax. Having both the old and new syntax, offering different behaviour still increases the overall complexity in the system.

I would rather expand upon the signal system if you are trying to do something that cannot be done or is too cumbersome to do, and change/improve the documentation.

**Comment by overhacked** [comment]
> `# if match then`
That is seriously a load-bearing comment. 😄 The lovely part to me about the rule system is that the regular expression matching is built in and is done in C, so it happens much faster than spawning multiple `grep` processes in a shell script from the signal system. It also is a nice, succinct DSL for attribute-based window properties instead of a bespoke shell script to maintain. That's why I'd prefer to handle it all in the rule system, since it is capable. Maybe adding `float=BOOL` as a rule property and documenting the difference between `manage` and `float`? (Which I understand much better now, thank you!)

I hear what you're saying about a new feature gated behind syntax. My thought was to expose the "evaluate all new rules at once" feature to people who would want to write their configuration file that way. Whether the rules get evaluated one-at-a-time or all together when they are created, though, is not that big of a deal since the most common case is that yabai starts during login before or at the same time as other windows. The sequential evaluation either doesn't matter (because there's no windows) or it results in a one-time shuffling of a few windows.

The case that I initially wanted to address that **does** impact my workflow is windows jumping around when they're created, especially when they reorder the layout because they jump into and out of floating. Merging the rules when they are applied to new windows eliminates that problem and doesn't change the existing behavior, because a "merged" set of window properties is the final state either way you do it.

The signal system seems to me to be a way to accomplish highly custom stuff not possible with the rules system. What I'm doing is 100% possible, and more concise, with the rules system but comes with an undesirable side-effect. I'd love to get a patch accepted to get rid of that side effect.

**Note:** I promise this is my last effort at convincing you. It's your project, so if you say, "Nope, not going to do it," after this I'll let it go. Thanks for engaging with me about the issue thus far!

**Comment by koekeishiya** [comment]
> The case that I initially wanted to address that does impact my workflow is windows jumping around when they're created, especially when they reorder the layout because they jump into and out of floating. Merging the rules when they are applied to new windows eliminates that problem and doesn't change the existing behavior, because a "merged" set of window properties is the final state either way you do it.

> The signal system seems to me to be a way to accomplish highly custom stuff not possible with the rules system. What I'm doing is 100% possible, and more concise, with the rules system but comes with an undesirable side-effect. I'd love to get a patch accepted to get rid of that side effect.

This is true. I'll give it some thought, but no promises.

**Comment by koekeishiya** [comment]
Implemented as part of #2123

**Comment by koekeishiya** [comment]
Fixed in v7.0.0.
